src/include/storage/record_buffer.h

```diff
@@ -32,22 +32,150 @@ class BufferSegment {
+     */
+    RecordType &operator*() const { return *reinterpret_cast<RecordType *>(segment_->bytes_ + segment_offset_); }
+
+    /**
+     * @return pointer to the underlying record
+     */
+    RecordType *operator->() const { return reinterpret_cast<RecordType *>(segment_->bytes_ + segment_offset_); }
+
+    /**
+     * prefix increment
+     * @return self-reference
+     */
+    Iterator &operator++() {
+      RecordType &me = this->operator*();
+      segment_offset_ += me.Size();
+      return *this;
+    }
+
+    /**
+     * postfix increment
+     * @return iterator that is equal to this before increment
+     */
+    const Iterator operator++(int) {
+      Iterator copy = *this;
+      operator++();
+      return copy;
+    }
+
+    /**
+     * equality check
+     * @param other the other iterator to compare to
+     * @return if the two iterators point to the same underlying record
+     */
+    bool operator==(const Iterator &other) const {
+      return segment_offset_ == other.segment_offset_ && segment_ == other.segment_;
+    }
+
+    /**
+     * inequality check
+     * @param other the other iterator to comapre to
+     * @return if the two iterators point to different underlying records
+     */
+    bool operator!=(const Iterator &other) const { return !(*this == other); }
+
+   private:
+    friend class IterableBufferSegment;
+    Iterator(BufferSegment *segment, uint32_t segment_offset) : segment_(segment), segment_offset_(segment_offset) {}
+    BufferSegment *segment_;
+    uint32_t segment_offset_;
+  };
+
+  /**
+   * Instantiates an IterableBufferSegment as a wrapper around a buffer segment
+   * @param segment
+   */
+  explicit IterableBufferSegment(BufferSegment *segment) : segment_(segment) {}
+
+  /**
+   * @return iterator to the first element
+   */
+  Iterator begin() { return {segment_, 0}; }
+
+  /**
+   * @return iterator to the second element
+   */
+  Iterator end() { return {segment_, segment_->size_}; }
+
+ private:
+  BufferSegment *segment_;
+};
+
+/**
+ * Custom allocator used for the object pool of buffer segments
+ */
+class RecordBufferSegmentAllocator {
+ public:
+  /**
+   * Allocates a new BufferSegment
+   * @return a new buffer segment
+   */
+  BufferSegment *New() {
+    auto *result = new BufferSegment;
+    TERRIER_ASSERT(reinterpret_cast<uintptr_t>(result) % 8 == 0, "buffer segments should be aligned to 8 bytes");
+    return result;
+  }
+
+  /**
+   * Resets the given buffer segment for use
+   * @param reused the buffer to reuse
+   */
+  void Reuse(BufferSegment *const reused) { reused->Reset(); }
+
+  /**
+   * Delete the given buffer segment and frees the memory
+   * @param ptr the buffer to delete
+   */
+  void Delete(BufferSegment *const ptr) { delete ptr; }
 };
 
+/**
+ * Type alias for an object pool handing out buffer segments
+ */
+using RecordBufferSegmentPool = common::ObjectPool<BufferSegment, RecordBufferSegmentAllocator>;
```

Comment:
Maybe just BufferSegmentPool?

Reply:
I'd rather be specific in case we have other buffers (network, e.g.).